File: Composite Pattern/LegoBauteil_cpp/LegoComposition.cpp
#include "LegoComposition.h"

LegoComposition::~LegoComposition()
{
	for (const auto piece : this->getPieces())
		delete piece;
}

LegoComposition::LegoComposition(std::vector<ILego*> pieces) : _pieces(std::move(pieces)) {}

LegoComposition::LegoComposition(const std::vector<ILego*>* pieces) : _pieces(*pieces) {}

auto LegoComposition::getPrice() const -> double
{
	double price = 0;
	for (const auto& piece : this->getPieces())
		if (piece != nullptr)
			price += piece->getPrice();
	return price;
}

auto LegoComposition::addPiece(ILego* p) -> void
{
	this->_pieces.emplace_back(p);
}

auto LegoComposition::removePiece(const ILego* p) const -> bool
{
	const auto pos = std::find(this->getPieces().begin(), this->getPieces().end(), p);
	if (pos != this->getPieces().end()) {
		this->getPieces().erase(pos);
		return true;
	}
	return false;
}

auto LegoComposition::getPieces() const -> std::vector<ILego*>
{
	return this->_pieces;
}

File: Composite/LegoBauteil_java/src/main/Lego/LegoComposition.java
package main.Lego;

import java.util.ArrayList;

public final class LegoComposition implements ILego {
    private final ArrayList<ILego> pieces;

    public LegoComposition() {
        this.pieces = new ArrayList<>();
    }

    public LegoComposition(final ArrayList<ILego> pieces) {
        this.pieces = pieces;
    }

    @Override
    public double getPrice() {
        double price = 0;
        for (var piece : pieces)
            price += piece.getPrice();
        return price;
    }

    public boolean addPiece(final ILego piece) {
        return this.getPieces().add(piece);
    }

    public boolean removePiece(final ILego piece) {
        return this.getPieces().remove(piece);
    }

    public ArrayList<ILego> getPieces() {
        return pieces;
    }
}
